src/main/include/log4cxx/private/mutex.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LOG4CXX_PRIV_MUTEX_HDR_
+#define LOG4CXX_PRIV_MUTEX_HDR_
+#include <mutex>
+#include <log4cxx/helpers/loglog.h>
+#include <log4cxx/helpers/stringhelper.h>
+#if !defined(LOG4CXX)
+	#define LOG4CXX 1
+#endif
+#include <log4cxx/private/log4cxx_private.h>
+
+namespace log4cxx
+{
+
+using AppenderMutexType = std::mutex;
+using AppenderScopeGuard = std::lock_guard<AppenderMutexType>;
+using AppenderScopedLock = std::unique_lock<AppenderMutexType>;
+
+// Change mutex attributes to avoid priority inversion problems
+template <class M>
+	void
+EnablePriorityInheritance(M& m, helpers::Pool& p)
```

Comment:
It looks like this doesn't do anything at the moment.  From the documentation that I've been reading, it seems that you can only set the mutexattr when creating the mutex in the first place(`pthread_mutex_init`).

Reply:
> From the documentation that I've been reading
Please post the link -  note that `pthread_mutex_destroy` does mean the it is being created.

Reply:
This was the best thing that I was able to find: https://linux.die.net/man/3/pthread_mutexattr_init

I think part of what's confusing me is that the multiple if statements are a strange way of writing the code, it didn't really occur to me that all of the if statements would be executed.

I suspect that the `pthread_mutex_destroy` and `pthread_mutex_init` on the native handle are technically undefined behavior though.  It probably won't cause issues now, but who knows what might change in the future?


Would a lockfree queue work, and/or changing the `std::mutex` to a `shared_mutex` work better than messing around with the internal mutex?  The original patch that was added in used effectively a `shared_mutex` with the lockfree queue(see: https://issues.apache.org/jira/browse/LOGCXX-500).  I think I removed everything that used `shared_mutex` to make it C++11 compatible, but there's still a compile test to see if it there(so we can easily switch between `std::shared_mutex` or `boost::shared_mutex`).

Reply:
> Would a lockfree queue work
No, the writes by an appender must be serialized.

> or changing the std::mutex to a shared_mutex
No, to prevent priority inversion, the low priority thread that holds the lock must have its priority raised (temporarily), which is an operating system thing.

Reply:
My understanding of this is as follows:
1. you have multiple threads logging to the same appender that have different priorities
2. that appender is NOT an `AsyncAppender`, so there is a priority inversion when a low-priority thread has IO control

If that is correct, I now understand better what this is doing.

Is there a reason that `AsyncAppender` is not appropriate in your situation?  It seems like that would be an easier way to solve the problem, since all of the IO then happens in a different thread.

Reply:
Using AsyncAppender could reduce the likelihood, but still result in a priority inversion. However, AsyncAppender would also increase context switching, which is not create on a real time controller.

A priority inversion problem is incredibily hard to verify and in my case only appears on deployed systems in the field. You cannot prove its absense with even extensive testing.

I believe it is an unfortunate accident of history the glibc does not do priority inheritance by default. I guess this is due to its relative recent discovery. Fortunately most applications do not have real time requirements. However the robotics community (which is a significant user of log4cxx - see ROS) really needs configurable logging and real time capabilities.

Reply:
Thanks for the explanation, that makes much more sense now.  I've never had to deal with a priority inversion before, so this has been a good learning experience.

One final thought: would it make more sense to create our own version of `std::mutex` that does the `pthread_mutexattr_setprotocol` in the constructor instead of effectively destroying and then re-initializing the underlying pthread mutex?